### ISSUE TITLE
Apply KL loss

### ISSUE THREAD

**Issue by weiqinchen7** [IDEA-FinAI/SQL-R1#9]
Hi,

Thank you for your contributions to the community!

When I tried to reproduce the experiments of SQL-R1, I had the issue that the **KL loss diverges to "Nan"**.

So I took a look at the codes (https://github.com/IDEA-FinAI/SQL-R1/blob/main/verl/trainer/ppo/ray_trainer.py#L665), and I saw below 

                        if **not** self.config.actor_rollout_ref.actor.use_kl_loss:
                            batch, kl_metrics = apply_kl_penalty(batch,
                                                                 kl_ctrl=self.kl_ctrl,
                                                                 kl_penalty=self.config.algorithm.kl_penalty)
                            metrics.update(kl_metrics)
                        else:
                            batch.batch['token_level_rewards'] = batch.batch['token_level_scores']

I wonder if the "not" in the codes above is a typo or we really need it? It looks like counterintuitive, and there is no "not" in the ray_trainer.py of the original VERL repo (https://github.com/volcengine/verl/blob/main/verl/trainer/ppo/ray_trainer.py#L1074) or see below

                        # compute rewards. apply_kl_penalty if available
                        if self.config.algorithm.use_kl_in_reward:
                            batch, kl_metrics = apply_kl_penalty(batch, kl_ctrl=self.kl_ctrl_in_reward, kl_penalty=self.config.algorithm.kl_penalty)
                            metrics.update(kl_metrics)
                        else:
                            batch.batch["token_level_rewards"] = batch.batch["token_level_scores"]

Please feel free to correct me if I am wrong. Thank you!

**Comment by MPX0222** [comment]
Thank you for your question! This issue seems to be caused by the update of the verl framework itself. You can refer to the following link https://github.com/volcengine/verl/pull/644

The framework of SQL-R1 has also been migrated from the older version. You can mainly focus on the reward and response length charts, which visually reflect the model's training situation.

**Comment by MPX0222** [comment]
This looks a bit strange. I haven't had such a situation during training. You can check the output of the checkpoints after Step 500 to see if there are garbled characters. This might indicate that the model training has crashed.

It should be noted that the parameters I published in A are relatively suitable for the training of the 7B model. If you need to train the 3B model, you need to adjust some parameters relatively, such as the learning rate, etc. I found that the 3B model does indeed require more precise adjustments, and this might also be one of the reasons for this situation

**Comment by weiqinchen7** [comment]
Thank you so much for your suggestions!

We currently fine-tune the original `Qwen2.5-Coder-3B-Instruct` model using the example data provided. But if I understand correctly, SQL-R1 paper proposes to first fine-tune the original `Qwen2.5-Coder-3B-Instruct` model by SFT over a 200K dataset and then fine-tune by GRPO over a 5K dataset. I think the `train.parquet` and `test.parquet` provided in the example_data folder is the 5K RL dataset right? And it seems like the 200K SFT dataset is not provided yet. 

Do you think this could be the reason? (that we skip the SFT and directly perform GRPO on the original `Qwen2.5-Coder-3B-Instruct` model). If so, is it possible that you can kindly open-source the SFT dataset as well? Thank you!

**Comment by MPX0222** [comment]
Yes, `train.parquet` and `test.parquet` is the 5K RL dataset.

I extracted the 200K data set directly from the original SynSQL-2.5D data set. Since I am in another region these months, you can first extract the data according to the method described in the paper for this experiment.

**Comment by MPX0222** [comment]
Oh, this is a key param for GRPO. It should be much larger, at least 4.

**Comment by weiqinchen7** [comment]
Agreed. Let me try `n=8` and see if it works. I'll keep you posted. Thank you!

**Comment by weiqinchen7** [comment]
Btw, I think I was wrong about the `not` question yesterday. The `not` in Line 665 of `verl/trainer/ppo/ray_trainer.py` is correct. It means that GRPO does not want to incorporate kl_penalty in the rewards, but rather adding it by a kl_loss. However, I couldn't find how the loss function is computed and where is the optimizer. Do you mind reminding me where it is? Appreciate it!

https://github.com/IDEA-FinAI/SQL-R1/blob/main/verl/trainer/ppo/ray_trainer.py#L665

**Comment by weiqinchen7** [comment]
I think I found it!

https://github.com/IDEA-FinAI/SQL-R1/blob/main/verl/workers/actor/dp_actor.py#L267

**Comment by MPX0222** [comment]
> I think I found it!
> 
> https://github.com/IDEA-FinAI/SQL-R1/blob/main/verl/workers/actor/dp_actor.py#L267

Agreed. This is the key line.
